packages/preferences/src/browser/views/preference-editor-widget.ts

```diff
@@ -169,21 +166,8 @@ export class PreferencesEditorWidget extends BaseWidget implements StatefulWidge
         }
     }
 
-    protected hasSchemaChanged(leafNode: Preference.LeafNode): boolean {
-        const oldKey = this.preferenceDataKeys.get(leafNode.id);
-        const newKey = this.setSchemaPropertyKey(leafNode);
-        return oldKey !== newKey;
-    }
-
-    protected setSchemaPropertyKey(leafNode: Preference.LeafNode): string | undefined {
-        const schemaProperty = this.schemaProvider.getSchemaProperty(leafNode.preferenceId);
-        if (schemaProperty) {
-            const key = stableJsonStringify(schemaProperty);
-            this.preferenceDataKeys.set(leafNode.id, key);
-            return key;
-        } else {
-            return undefined;
-        }
+    protected hasSchemaChanged(renderer: GeneralPreferenceNodeRenderer, node: Preference.LeafNode): boolean {
+        return !deepEqual(renderer.schema, node.preference.data);
```

Comment:
I don't think this `deepEqual` call ever returns `false`, since `renderer.schema` and `node.preference.data` always have the same object pointer. That's the reason why I needed to cache the stringified object in the first place.

Reply:
Hm... that's a fair point. In the case I was most interested in, I was putting a new object in place, so the old version and the new version weren't strictly equal, but you're right, if only the internals are changed, that wouldn't be the case.

Reply:
On the other hand, that's really in the hands of the updater. By default, the schema provider just assigns the object passed into to `updateSchemaProperty` to the appropriate key, so if a new object is passed in, then there won't be referential equality:

```
updateSchemaProperty(key: string, property: PreferenceDataProperty): void {
        this.updateSchemaProps(key, property);
        this.fireDidPreferenceSchemaChanged();
    }

    protected updateSchemaProps(key: string, property: PreferenceDataProperty): void {
        this.combinedSchema.properties[key] = property;

        switch (property.scope) {
            case PreferenceScope.Folder:
                this.folderSchema.properties[key] = property;
            // Fall through. isValidInScope implies that User ⊃ Workspace ⊃ Folder,
            // so anything we add to folder should be added to workspace, but not vice versa.
            case PreferenceScope.Workspace:
                this.workspaceSchema.properties[key] = property;
                break;
        }
    }
```

Reply:
Pushed a change that adjusts the existing uses of `updateSchemaProperty` to use new objects, and added a comment on the method to indicate that that's the recommended usage. What do you think of that approach?